### PR TITLE
ALL the code until now :)

### PR DESCRIPTION
**Pull request by pjazdzewski1990**

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,12 @@
+root = true
```
Comment (pjazdzewski1990) [inline .editorconfig:1]: Isn't this like private for people?

File: src/components/App/App.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import ProductDetails from '../../screens/ProductDetails/ProductDetails'
+
+//let yeomanImage = require('../../images/yeoman.png');
+
+const App = () => {
+  return (
+    <div className="index">
+      <Navbar/>
+
+      <Switch>
```
Comment (pjazdzewski1990) [inline src/components/App/App.js]: What is the difference between `<Switch>` and `<Router>`?

File: src/components/ProductsList/ListElement/ListElementStyle.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+// language=SCSS prefix=dummy{ suffix=}
+const ListElement = styled.li`
+    font-size: 20px;
+    margin: 20px 0;
+    padding: 20px;
+    border: black 2px solid;
+    height: 100px;
+    
+    &:hover {
```
Comment (pjazdzewski1990) [inline src/components/ProductsList/ListElement/ListElementStyle.js]: Not a problem with this particular code, but I wonder ... doesn't this approach to define styled components prevent reuse? Or on the contrary - it's designed to prevent reuse?

File: src/data/products.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    'Fusce dapibus posuere condimentum. Maecenas malesuada lorem magna, quis ultrices nunc laoreet sit amet. ' +
+    'Phasellus vitae erat lacus. Morbi non eros eget nibh pretium convallis tristique et lacus. '
+  },
+  {
+    id:5,
+    name:'product5',
+    price:17,
+    category:2
+  }
+]
```
Comment (pjazdzewski1990) [inline src/data/products.js]: If this is defined in code I would push it further and generated products in loop. Many products, say ~500, will pose more challenge and  reveal issues we might be missing now

File: src/screens/Shop/Shop.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+import Screen from '../../styledComponents/Screen'
+import Filters from '../../components/Filters/Filters'
+import ProductsList from '../../components/ProductsList/ProductsList'
+
+const Shop = () => {
+    return (
+      <Screen>
+        <Filters/>
+        <ProductsList/>
```
Comment (pjazdzewski1990) [inline src/screens/Shop/Shop.js]: How does `ProductsList` know what to show?

File: src/screens/ShoppingCart/ShoppingCart.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        </ShoppingCartList>
+
+        <Link to='/'>
+          <Button>Back to shop</Button>
+        </Link>
+      </Screen>
+    )
+}
+
+export default connect(mapStateTopProps,mapDispatchToProps)(ShoppingCart)
```
Comment (pjazdzewski1990) [inline src/screens/ShoppingCart/ShoppingCart.js]: I think you can  use decorators you make this automatic - I think Learn-Js repo uses it

File: src/components/App/App.js (hunk 2 above)
Comment (Alwox) [reply src/components/App/App.js]: I'm using here also `<Router>` but in new version form: `<BrowserRouter>` (old `<Router history={browserHistory}>`) and in index.js file.

### Switch
Renders the first child `<Route>` or `<Redirect>` that matches the location.

### Router
The common low-level interface for all router components. Typically apps will use one of the high-level routers instead: `<BrowserRouter>`,[...].The most common use-case for using the low-level `<Router> `is to synchronize a custom history with a state management lib like Redux or Mobx.

File: src/data/products.js (hunk 4 above)
Comment (Alwox) [reply src/data/products.js]: Yes, it's on a to-do list. I was thinking about more 'real' products list in future but maybe it's better to generate it now and possibly change later.

File: src/components/ProductsList/ListElement/ListElementStyle.js (hunk 3 above)
Comment (Alwox) [reply src/components/ProductsList/ListElement/ListElementStyle.js]: You mean height? I know it's bad approach also because of responsiveness but all styles are to rewrite. I was doing it fastest way because at the beginning I haven't plan how it could look ;)

File: src/screens/Shop/Shop.js (hunk 5 above)
Comment (Alwox) [reply src/screens/Shop/Shop.js]: It imports all products from a file.

File: .editorconfig (hunk 1 above)
Comment (Alwox) [reply .editorconfig:1]: I'm not sure what you mean.

> `root = true`
This line tells EditorConfig this is the root of the project, thus EditorConfig won’t apply the settings of .editorconfig outside this directory.

Look like it's ok.

File: src/components/App/App.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+require('normalize.css/normalize.css')
+require('./App.css')
+
+import React from 'react'
+import Navbar from '../Navbar/Navbar'
```
Comment (pmusiol) [inline src/components/App/App.js]: Maybe for more clarity we could move App.js out of components dir? It's the most generic, would make imports here slightly nicer, plus it's not widely used like the other components.

File: src/screens/ShoppingCart/ShoppingCart.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    const price =
+      productsData.find(
+        prod => prod.id === parseInt(product)
+      ).price
+
+    sum += price * products[product]
+
+    productsTable.push([name,price,products[product],parseInt(product)])
+  }
+
```
Comment (piecyk) [inline src/screens/ShoppingCart/ShoppingCart.js]: hmm maybe you can us here some lodash, ramda js helpers 
( IMHO hard to read the code ) 

something around ?
```js

const {productsInCart, sum} = _.reduce(
  _.groupBy(props.productsInCart, 'id'), 
  (acc, products) => {
    const product = _.head(products);
    const count = _.size(product);
    return {
      sum: acc.sum + (count * product.price),
      productsInCart: [...acc.productsInCart, {...product, count}]
    }
  }, 
  {productsInCart: []: sum: 0}
);
```

File: src/screens/ShoppingCart/ShoppingCart.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      ).name
+
+    const price =
+      productsData.find(
+        prod => prod.id === parseInt(product)
+      ).price
+
+    sum += price * products[product]
+
+    productsTable.push([name,price,products[product],parseInt(product)])
```
Comment (piecyk) [inline src/screens/ShoppingCart/ShoppingCart.js]: IMO better to use here object `product` just extend it to have `count`

File: src/components/Filters/Filters.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  return {
+    setProductsSort: (sortType) => dispatch(setProductsSort(sortType))
+  }
+}
+const Filters = (props) => {
+    return (
+      <FilterStyle>
+        <span>Sort by price:</span>
+
+        <Button active={props.sortType === 'PRICE_ASC' ? 'active' : false}
```
Comment (pmusiol) [inline src/components/Filters/Filters.js]: How about calling the <Button> with some less generic name? It's not very obvious it's your component, not basic html <button>

File: src/styledComponents/Button.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import styled from 'styled-components'
+
+// language=SCSS prefix=dummy{ suffix=}
+const Button = styled.button`
+    padding: 6px 10px;
+    margin: 6px;
+    border: 1px solid white;
+    background-color: transparent;
+    font-size: 16px;
+    box-shadow: ${props => props.active ? 'inset 0 0 4px 1px #848484' : 'none'};
```
Comment (pmusiol) [inline src/styledComponents/Button.js]: Maybe we can call that props different, not to be confused with :active state?

File: package.json [hunk 12]
```diff
@@ -64,6 +65,10 @@
     "core-js": "^2.0.0",
     "normalize.css": "^4.0.0",
     "react": "^15.0.0",
-    "react-dom": "^15.0.0"
+    "react-dom": "^15.0.0",
+    "react-redux": "^5.0.4",
+    "react-router-dom": "^4.1.1",
+    "redux": "^3.6.0",
+    "styled-components": "^1.4.6"
```
Comment (piecyk) [inline package.json:44]: Didn't use `styled-components` but don't think it will work in long run... I think better way is to use `namespaces` in stylus and some loader to load it automatically (if you don't want to import them implicit [the stylus files] ) 

for example  
```stylus
// MyA.styl
#ns(my-a)
   color #ddd
   .title
      font-size 16px
```
```jsx
// MyA.jsx
import namespaced from 'namespaced'; // https://github.com/hvesalai/namespaced

const cln = namespaced('my-a');

const MyA = ({children}) => (
  <div className={cln(':')}>
     <div className={cln(':title')}>MyA</div>
     {children}
  </div>
);

export default MyA;

```
IMHO this more clean, easy to read and maintain,

File: src/screens/ProductDetails/ProductDetails.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  return {
+    addToCart: (product,amount) => dispatch(addToCart(product,amount))
+  }
+}
+
+const ProductDetails = (props) => {
+
+  const id = parseInt(props.match.params.id, 10)
+  const product = products.find(product => product.id === id)
+  const category = categories.find(category => category.id === product.category).name
```
Comment (pmusiol) [inline src/screens/ProductDetails/ProductDetails.js]: Very very small issue, but maybe for a sake of symmetry with const product, category should remain the whole item, not .name?

File: src/screens/ProductDetails/ProductDetails.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+import Button from '../../styledComponents/Button'
+import products from '../../data/products'
+import categories from '../../data/categories'
+import {connect} from 'react-redux'
+import { Link } from 'react-router-dom'
+import {addToCart} from '../../screens/ShoppingCart/actionCreators'
+
+const mapDispatchToProps = (dispatch) => {
+  return {
+    addToCart: (product,amount) => dispatch(addToCart(product,amount))
```
Comment (pmusiol) [inline src/screens/ProductDetails/ProductDetails.js]: It seems, that amount isn't really supported, right?
If, for some reasons, it has to be here, it would at least set it with default value =1.

File: src/components/App/App.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+import { Switch, Route } from 'react-router-dom'
+import Shop from '../../screens/Shop/Shop'
+import ShoppingCart from '../../screens/ShoppingCart/ShoppingCart'
+import ProductDetails from '../../screens/ProductDetails/ProductDetails'
```
Comment (piecyk) [inline src/components/App/App.js]: Relative paths can be hard to follow / maintain in larger projects. We can use absolute paths (relative to project `src`) adding `webpack` config something like

`resolve: { modules: [path.resolve('./src'), etc..]}`
  
then you could write 
`import ProductDetails from 'screens/ProductDetails/ProductDetails'`

File: src/components/App/App.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+require('normalize.css/normalize.css')
+require('./App.css')
+
+import React from 'react'
+import Navbar from '../Navbar/Navbar'
+import { Switch, Route } from 'react-router-dom'
+import Shop from '../../screens/Shop/Shop'
```
Comment (piecyk) [inline src/components/App/App.js]: Are in `redux` world `screens` called `containers` ?

File: src/components/App/App.js (hunk 7 above)
Comment (Alwox) [reply src/components/App/App.js]: I moved it to components because there are 6 folders in src now so I was thinking maybe is better to limit it for readability especially if this one is used quite rarely. But yes, now I see it wasn't best idea and anyway this project folders structure need some improvement ;)

File: src/components/Filters/Filters.js (hunk 10 above)
Comment (Alwox) [reply src/components/Filters/Filters.js]: I think capitalize is enough to see the difference (especially if it's general React rule) and complexity can be confusing if it's still just button. For example, react-bootstrap is using the same naming.

File: .editorconfig (hunk 1 above)
Comment (pjazdzewski1990) [reply .editorconfig:1]: What I meant - should this file be part of the repository or should it be owned by a developer (hence `.gitignore`d)

File: src/styledComponents/Button.js (hunk 11 above)
Comment (Alwox) [reply src/styledComponents/Button.js]: I will change it to `isActive`.

File: src/components/ProductsList/ListElement/ListElementStyle.js (hunk 3 above)
Comment (pjazdzewski1990) [reply src/components/ProductsList/ListElement/ListElementStyle.js]: I'm talking in general

File: src/screens/Shop/Shop.js (hunk 5 above)
Comment (pjazdzewski1990) [reply src/screens/Shop/Shop.js]: Is this a good practice? Importing from file is ok, but we should at least pretend that some kind of server is the source

File: src/screens/ProductDetails/ProductDetails.js (hunk 14 above)
Comment (Alwox) [reply src/screens/ProductDetails/ProductDetails.js]: I made it because I was thinking I will use it right after but then this feature goes for 'later' ;) I will use it in next update.

File: src/screens/ProductDetails/ProductDetails.js (hunk 13 above)
Comment (Alwox) [reply src/screens/ProductDetails/ProductDetails.js]: Maybe but `category` in fact is just a name (ok, and id) so it can also be a bit confusing 'do we have there something else?'.

File: src/components/App/App.js (hunk 15 above)
Comment (Alwox) [reply src/components/App/App.js]: Thanks, it's really helpful ;)

File: src/components/App/App.js (hunk 16 above)
Comment (Alwox) [reply src/components/App/App.js]: I'm not sure what you mean?

Comment (piecyk) [reply src/components/App/App.js]: While browsing `redux` examples they use `containers` as name for root page components 
just like the name better than `screens` 💃 

https://github.com/reactjs/redux/tree/master/examples/real-world/src/containers

File: .editorconfig (hunk 1 above)
Comment (Alwox) [reply .editorconfig:1]: In fact I'm not sure.

File: src/screens/Shop/Shop.js (hunk 5 above)
Comment (Alwox) [reply src/screens/Shop/Shop.js]: I want to add in future async loading for this data.

File: src/screens/ShoppingCart/ShoppingCart.js (hunk 8 above)
Comment (Alwox) [reply src/screens/ShoppingCart/ShoppingCart.js]: I think I can try to add lodash to project.

File: src/components/App/App.js (hunk 16 above)
Comment (Alwox) [reply src/components/App/App.js]: I was using this article as example https://medium.com/@alexmngn/how-to-better-organize-your-react-applications-2fd3ea1920f1 (i have no idea how scenes become screens 😅  ) for me it sounds more natural because container look like something just to keep other components but here it also have own elements (but maybe it's bad?).

Comment (piecyk) [reply src/components/App/App.js]: hmm i will read that later, just from quick look i think it's bad idea the have two folders for components, like `components` and `styledComponents` merge them together 
also `data, images` will move to `lib`, drop `config` it should be all provided as env from webpack at build time.

hmm i think containers i also good name here ( it contains all child components needed for page/screen to be created ) my comments was just about that `containers` was more used name and can be more known ( but really not important ) at this stage 

also next i would update to webpack2 and clean the config stuff there

Comment (Alwox) [reply src/components/App/App.js]: Yes, I wanted to reorganize folders structure now - in fact, all components in `components` folder are used once and I think it's better moved them to `containers` which are using them. Then we end with just one `components` folder anyway.

> `data, images` will move to `lib`

you mean to `dist` folder? Or ceate new `lib`?

Comment (piecyk) [reply src/components/App/App.js]: yeah, the rules should be if something is only needed for container component it goes there but some more generic extract to components even if it's only used one time in time of creation 

i would recommend something around this lines ...
( also when you go with absolute paths then it's easy to move stuff around  )
```
 ./src
   ./lib
      ./data
      ./images
      ./services ( some common services for api handling )
      ./utils ( utils needed cross components/containers like DomUtil etc... )
   ./components
      ./Button
         Button.jsx
   ./containers 
      ./home
         Home.jsx  
```
Not to `dist`, this folder should be removed on every build, and re created from start and shuld be on root level of repo as `./src`

Comment (Alwox) [reply src/components/App/App.js]: Thanks 😉

Comment (Alwox) [reply src/components/App/App.js]: I met some problem - I want divide components into presentational and containers (https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) but now `containers` folder name can be confusing. I don't think to put there all my containers is a good idea. I want to keep there on the top only my 'views' so maybe is better stay with `scenes` or `screens` naming?

Comment (piecyk) [reply src/components/App/App.js]: hmm that's good idea, but as with everything don't go all the way down the rabbit hole...

Basic don't create `Container Component` (or as you want to named them) for everything, there's no need for some smaller ones just not need noise and it will grow fast!  

I'm calling `container` components `Controllers` (form old times it just stick) so for example you have `Home.jsx` ( page |  container | scene | screen ) that fetch some data and pass it down to
`HomeController.jsx` that mange actions, maybe have some local state, here you include your components and build layout, if this get to complex you can extract from it `HomeLayout.jsx` that will be sliced more etc... you got the idea... it all depends on complexity of layout you are building

Comment (Alwox) [reply src/components/App/App.js]: Ok, thank you a lot 😉 I'll read and think more about this and we will see results soon 🙂

Comment (piecyk) [reply src/components/App/App.js]: yeah no problem man, ping me if you are stuck on something 👍 awesome work 👏

File: src/App/App.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
+require('normalize.css/normalize.css')
+require('./App.css')
```
Comment (piecyk) [inline src/App/App.js]: Why here you use `require` ? Don't mix them, use `import` in all places

File: src/App/App.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+        <Route exact path='/' component={Shop}/>
+        <Route path='/cart' component={ShoppingCart}/>
+        <Route path='/:id' component={ProductDetails}/>
+      </Switch>
+
+    </div>
+  )
+}
+
+export default App
```
Comment (piecyk) [inline src/App/App.js]: I would move App to `containers` as it is a root container, also you can go with more compact construction in arrow function declaration

 ```jsx

const AppLayout = ({children}) => (
  <div className='index'>
    <Navbar/>
    {children}
  </div>
);

const App = () => (
  <AppLayout>
    <Switch>
      <Route exact path='/' component={Shop}/>
      <Route path='/cart' component={ShoppingCart}/>
      <Route path='/:id' component={ProductDetails}/>
    </Switch>    
  </AppLayout>  
);

export default App;
```

File: src/App/App.js (hunk 17 above)
Comment (Alwox) [reply src/App/App.js]: This file was generated by Yeoman generator and I forget to change it, thanks. But I don't know why they made this mix there.

File: src/containers/ProductDetails/ProductDetailsView.js [hunk 19]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import Screen from '../../components/Screen'
+import Button from '../../components/Button'
+import { Link } from 'react-router-dom'
+
+const ProductDetailsView = (props) => {
```
Comment (piecyk) [inline src/containers/ProductDetails/ProductDetailsView.js]: The same here with `Object destructuring` it get more compact and in quick look you can figure out props used 
```jsx
const ProductDetailsView = ({product, addToCart}) => (
  <Screen>
```

File: src/App/App.js (hunk 18 above)
Comment (Alwox) [reply src/App/App.js]: Thanks, I forgot that I don't need `return` if it's only JSX but I'm not sure what is profit in divide this into two components?

Comment (piecyk) [reply src/App/App.js]: Here just as example, there is no real benefit... In some more complex components i try to extract chunks of dom to be more easy to read (simple stateless presentational components)
